sync/src/net_time_checker.rs

```diff
@@ -0,0 +1,165 @@
+pub struct NetTimeChecker {
+    /// Local clock should has less offset than this value.
+    tolerant_offset: u64,
+    max_samples: usize,
+    min_samples: usize,
+    samples: VecDeque<i64>,
+}
+
+impl NetTimeChecker {
+    pub fn new(min_samples: usize, max_samples: usize, tolerant_offset: u64) -> Self {
+        NetTimeChecker {
+            min_samples,
+            max_samples,
+            tolerant_offset,
+            samples: VecDeque::with_capacity(max_samples + 1),
+        }
+    }
+
+    pub fn add_sample(&mut self, offset: i64) {
+        self.samples.push_back(offset);
+        if self.samples.len() > self.max_samples {
+            self.samples.pop_front();
+        }
+    }
+
+    fn median_offset(&self) -> Option<i64> {
+        if self.samples.is_empty() || self.samples.len() < self.min_samples {
+            return None;
+        }
+        let mut samples = self.samples.iter().cloned().collect::<Vec<_>>();
+        samples.sort();
+        if samples.len() % 2 == 0 {
+            // samples is even
+            let i = samples.len() / 2;
+            Some((samples[i - 1] + samples[i]) / 2)
+        } else {
+            // samples is odd
+            samples.get(samples.len() / 2).cloned()
+        }
+    }
+
+    pub fn check(&self) -> Result<(), i64> {
+        let network_offset = match self.median_offset() {
+            Some(offset) => offset,
+            None => return Ok(()),
+        };
+        if network_offset.abs() as u64 > self.tolerant_offset {
+            return Err(network_offset);
+        }
+        Ok(())
+    }
+}
+
+impl Default for NetTimeChecker {
+    fn default() -> Self {
+        NetTimeChecker::new(MIN_SAMPLES, MAX_SAMPLES, TOLERANT_OFFSET)
+    }
+}
+
+/// Collect time offset samples from network peers and send notify to user if offset is too large
+pub struct NetTimeProtocol(RwLock<NetTimeChecker>);
+
+impl NetTimeProtocol {
+    pub fn new(min_samples: usize, max_samples: usize, tolerant_offset: u64) -> Self {
+        NetTimeProtocol(RwLock::new(NetTimeChecker::new(
+            min_samples,
+            max_samples,
+            tolerant_offset,
+        )))
+    }
+}
+
+impl Default for NetTimeProtocol {
+    fn default() -> Self {
+        NetTimeProtocol(RwLock::new(NetTimeChecker::default()))
+    }
+}
+
+impl CKBProtocolHandler for NetTimeProtocol {
+    fn initialize(&self, _nc: Box<CKBProtocolContext>) {}
+    fn received(&self, nc: Box<CKBProtocolContext>, peer: PeerIndex, data: &[u8]) {
+        // collect time sample from outbound peer
+        if nc.session_info(peer).map(|s| s.peer.is_outbound()) == Some(true) {
+            let now = faketime::unix_time_as_millis();
+            let timestamp = match get_root::<TimeMessage>(&data)
+                .payload()
+                .map(|time| time.timestamp())
+            {
+                Some(timestamp) => timestamp,
+                None => return,
+            };
+            let offset: i64 = if now > timestamp {
```

Comment:
```suggestion
            let offset = (now  as i64) - (timestamp as i64);
```

Reply:
`now` is u64, direct cast it to i64 may cause overflow, instead, we cast the calculation result.

Reply:
`overflowing_sub`